File: c++/src/xdevs/examples/efp/Ef.cpp
/*
 * Copyright (C) 2016-2016 José Luis Risco Martín <jlrisco@ucm.es>.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see
 * http://www.gnu.org/licenses/
 *
 * Contributors:
 *  - José Luis Risco Martín
 */

#include "Ef.h"

Ef::Ef(const std::string& name, const double& period, const double& observationTime)
			: Coupled(name),
			  generator("generator", period),
			  transducer("transducer", observationTime),
			  iIn("in"),
			  oOut("out") {
	Component::addInPort(&iIn);
	Component::addOutPort(&oOut);
	Coupled::addComponent(&generator);
	Coupled::addComponent(&transducer);
	Coupled::addCoupling(this, &(this->iIn), &transducer, &transducer.iSolved);
	Coupled::addCoupling(&generator, &generator.oOut, this, &(this->oOut));
	Coupled::addCoupling(&generator, &generator.oOut, &transducer, &transducer.iArrived);
	Coupled::addCoupling(&transducer, &transducer.oOut, &generator, &generator.iStop);
}

Ef::~Ef() { }

File: src/xdevs/core/examples/efp/Ef.java
/*
 * Copyright (C) 2014-2015 José Luis Risco Martín <jlrisco@ucm.es> and 
 * Saurabh Mittal <smittal@duniptech.com>.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see
 * http://www.gnu.org/licenses/
 *
 * Contributors:
 *  - José Luis Risco Martín
 */
package xdevs.core.examples.efp;

import xdevs.core.modeling.Coupled;
import xdevs.core.modeling.Port;

/**
 *
 * @author José Luis Risco Martín
 */
public class Ef extends Coupled {

  protected Port<Job> iStart = new Port<>("iStart");
  protected Port<Job> iIn = new Port<>("iIn");
  protected Port<Job> oOut = new Port<>("oOut");

  public Ef(String name, double period, double observationTime) {
	  super(name);
    super.addInPort(iIn);
    super.addInPort(iStart);
    super.addOutPort(oOut);
    Generator generator = new Generator("generator", period);
    super.addComponent(generator);
    Transducer transducer = new Transducer("transducer", observationTime);
    super.addComponent(transducer);
    
    super.addCoupling(this.iIn, transducer.iSolved);
    super.addCoupling(generator.oOut, this.oOut);
    super.addCoupling(generator.oOut, transducer.iArrived);
    super.addCoupling(transducer.oOut, generator.iStop);
    super.addCoupling(this.iStart, generator.iStart);
  }
}
